pypose/lietensor/operation.py

```diff
@@ -328,4 +457,291 @@ def backward(ctx, grad_output):
+        m = SO3_Matrix(X)
+        zero = torch.zeros(X.shape[:-1]+(1,), device=X.device, dtype=X.dtype)
+        X_grad = dq @ SO3_Act_Jacobian(out)
+        p_grad = dq @ m[..., :3, :3]
+        return torch.cat((X_grad.squeeze(-2), zero), dim = -1), p_grad.squeeze(-2)
+
+
+class SE3_Act(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, X, p):
+        out = X[..., :3] + SO3_Act.apply(X[..., 3:], p)
+        ctx.save_for_backward(X, out)
+        return out
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        X, out = ctx.saved_tensors
+        dq = grad_output[..., :3].unsqueeze(-2)
+        m = SE3_Matrix(X)
+        zero = torch.zeros(X.shape[:-1]+(1,), device=X.device, dtype=X.dtype)
+        X_grad = dq @ SE3_Act_Jacobian(out)
+        p_grad = dq @ m[..., :3, :3]
+        return torch.cat((X_grad.squeeze(-2), zero), dim = -1), p_grad.squeeze(-2)
+
+
+class RxSO3_Act(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, X, p):
+        out = X[..., 4:] * SO3_Act.apply(X[..., :4], p)
+        ctx.save_for_backward(X, out)
+        return out
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        X, out = ctx.saved_tensors
+        dq = grad_output[..., :3].unsqueeze(-2)
+        m = RxSO3_Matrix(X)
+        zero = torch.zeros(X.shape[:-1]+(1,), device=X.device, dtype=X.dtype)
+        X_grad = dq @ RxSO3_Act_Jacobian(out)
+        p_grad = dq @ m[..., :3, :3]
+        return torch.cat((X_grad.squeeze(-2), zero), dim = -1), p_grad.squeeze(-2)
+
+
+class Sim3_Act(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, X, p):
+        out = X[..., :3] + RxSO3_Act.apply(X[..., 3:], p)
+        ctx.save_for_backward(X, out)
+        return out
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        X, out = ctx.saved_tensors
+        dq = grad_output[..., :3].unsqueeze(-2)
+        m = Sim3_Matrix(X)
+        zero = torch.zeros(X.shape[:-1]+(1,), device=X.device, dtype=X.dtype)
+        X_grad = dq @ Sim3_Act_Jacobian(out)
+        p_grad = dq @ m[..., :3, :3]
+        return torch.cat((X_grad.squeeze(-2), zero), dim = -1), p_grad.squeeze(-2)
+
+
+class SO3_Act4(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, X, p):
+        out = torch.cat((SO3_Act.apply(X, p[..., :3]), p[..., 3:]), dim=-1)
+        ctx.save_for_backward(X, out)
+        return out
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        X, out = ctx.saved_tensors
+        dq = grad_output.unsqueeze(-2)
+        zero = torch.zeros(X.shape[:-1]+(1,), device=X.device, dtype=X.dtype)
+        X_grad = dq @ SO3_Act4_Jacobian(out)
+        p_grad = dq @ SO3_Matrix4x4(X)
+        return torch.cat((X_grad.squeeze(-2), zero), dim = -1), p_grad.squeeze(-2)
+
+
+class SE3_Act4(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, X, p):
+        out = torch.cat(((SO3_Act.apply(X[..., 3:], p[..., :3]) + X[..., :3] * p[..., 3:]), p[..., 3:]), dim=-1)
```

Comment:
Too long, how about splitting into two lines, e.g.,
```
name = O3_Act.apply(X[..., 3:], p[..., :3]) + X[..., :3] * p[..., 3:]
out = torch.cat((name , p[..., 3:]), dim=-1)
```
or
```
name = O3_Act.apply(X[..., 3:], p[..., :3]) 
out = torch.cat((name + X[..., :3] * p[..., 3:], p[..., 3:]), dim=-1)
```
remember to give a reasonable `name` instead.